Branch-SDK/Branch-SDK/BNCServerInterface.m

```diff
@@ -11,109 +11,130 @@
+        return [self preparePostRequest:post url:url retryNumber:++lastRetryNumber log:log];
+    }];
 }
 
-- (BNCServerResponse *)postRequestSync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData log:(BOOL)log {
+- (BNCServerResponse *)postRequest:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData log:(BOOL)log {
     NSURLRequest *request = [self preparePostRequest:post url:url retryNumber:0 log:log];
-    return [self genericSyncHTTPRequest:request withTag:requestTag andLinkData:linkData];
+    return [self genericHTTPRequest:request withTag:requestTag andLinkData:linkData];
 }
 
-- (void)genericAsyncHTTPRequest:(NSMutableURLRequest *)request withTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData {
-    [NSURLConnection sendAsynchronousRequest:request queue:[NSOperationQueue mainQueue] completionHandler: ^(NSURLResponse *response, NSData *POSTReply, NSError *error) {
-        BNCServerResponse *serverResponse = [self processServerResponse:response data:POSTReply error:error tag:requestTag andLinkData:linkData];
-        if (self.delegate) [self.delegate serverCallback:serverResponse];
+
+#pragma mark - Generic requests
+
+- (void)genericHTTPRequest:(NSURLRequest *)request withTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData callback:(BNCServerCallback)callback {
+    [self genericHTTPRequest:request withTag:requestTag andLinkData:linkData retryNumber:0 log:YES callback:callback retryHandler:^NSURLRequest *(NSInteger lastRetryNumber) {
+        return request;
     }];
 }
 
-- (BNCServerResponse *)genericSyncHTTPRequest:(NSURLRequest *)request withTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData {
-    NSURLResponse * response = nil;
-    NSError * error = nil;
-    NSData * POSTReply = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];
+- (void)genericHTTPRequest:(NSURLRequest *)request withTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData retryNumber:(NSInteger)retryNumber log:(BOOL)log callback:(BNCServerCallback)callback retryHandler:(NSURLRequest *(^)(NSInteger))retryHandler {
+    [NSURLConnection sendAsynchronousRequest:request queue:[NSOperationQueue mainQueue] completionHandler:^(NSURLResponse *response, NSData *responseData, NSError *error) {
+        BNCServerResponse *serverResponse = [self processServerResponse:response data:responseData error:error tag:requestTag andLinkData:linkData];
+        NSInteger status = [serverResponse.statusCode integerValue];
+        BOOL isRetryableStatusCode = status >= 500;
+        
+        // Retry the request if appropriate
+        if (retryNumber < [BNCPreferenceHelper getRetryCount] && isRetryableStatusCode) {
```

Comment:
This is called an intention variable (struggling to find a good link), and is used for the sake of readability. If I was to wander through this code base and find a `&& status > 500`, it wouldn't really be clear to me why a status greater than 500 should cause this check to pass. In this case, we're basically defining statuses greater than 500 as retryable statuses.